components/schemes/Shape/src/MDCShapeCorner.h

```diff
@@ -0,0 +1,91 @@
+// Copyright 2018-present the Material Components for iOS authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import <Foundation/Foundation.h>
+#import "MaterialShapeLibrary.h"
+
+/**
+ This enum consists of the different types of shape corners.
+
+ - MDCShapeCornerFamilyRounded: A rounded corner.
+ - MDCShapeCornerFamilyAngled: An angled/cut corner.
+ */
+typedef NS_ENUM(NSInteger, MDCShapeCornerFamily) {
+  MDCShapeCornerFamilyRounded,
+  MDCShapeCornerFamilyAngled,
+};
+
+/**
+ This enum consists of the different types of shape values that can be provided.
+
+ - MDCShapeCornerSizeTypeAbsolute: If an absolute shape size is provided.
+ - MDCShapeCornerSizeTypePercentage: If a relative shape size is provided.
+ */
+typedef NS_ENUM(NSInteger, MDCShapeCornerSizeType) {
+  MDCShapeCornerSizeTypeAbsolute,
+  MDCShapeCornerSizeTypePercentage,
+};
+
+/**
+ An MDCShapeCorner is the shape value of a corner. It takes a family, a size type, and a size.
+ */
+@interface MDCShapeCorner : NSObject
+
+/**
+ The shape family of our corner.
+ */
+@property(assign, nonatomic) MDCShapeCornerFamily family;
+
+/**
+ The size type of our shape.
+ */
+@property(assign, nonatomic) MDCShapeCornerSizeType sizeType;
+
+/**
+ The size of our shape.
+
+ When MDCShapeSizeType is MDCShapeCornerSizeTypeAbsolute, this accepts absolute values
+ that are positive.
+
+ When MDCShapeSizeType is MDCShapeCornerSizeTypePercentage, this accepts percentage values
+ from 0 to 1 (0% - 100%). These values are percentages based on the height of the surface.
+ */
+@property(assign, nonatomic) CGFloat size;
+
+/**
+ This initializer takes in a shape family and size, and returns a shape corner.
+
+ @param cornerFamily The shape family.
+ @param cornerSize The shape size.
+ @return an MDCShapeCorner.
+ */
+- (instancetype)initWithFamily:(MDCShapeCornerFamily)cornerFamily andSize:(CGFloat)cornerSize;
+
+/**
+ This method returns an MDCCornerTreament representation of the MDCShapeCorner.
+
+ @return an MDCCornerTreatment.
+ */
+- (MDCCornerTreatment *)cornerTreatmentValue;
```

Comment:
Why do we have this layer of indirection? Do we intend to deprecate or replace MDCCornerTreatment?

Reply:
Put another way - why is the logic that turns a shape corner into a corner treatment not part of the themers instead?

Reply:
I tried to make it so we only have to do the conversion of `MDCShapeCorner` to one of the `MDCCornerTreatment` instances once, rather than it being done in every one of our themers. I initially had it in the themers but it was rather messy and redundant. In terms of architecture it made sense to add a convenience converter in `MDCShapeCorner` and then add additional initializers to `MDCCornerTreatment` to take the needed information and initialize the right corner treatment. Do you think there is a better way?

Reply:
> Do we intend to deprecate or replace MDCCornerTreatment?

What are your thoughts on this question?

Reply:
I think `MDCCornerTreatment` fits the shape subsystem's purpose of setting corner treatments inside an `MDCRectangleShapeGenerator` well. I also think that the `MDCShapeCorner` encapsulates the exact relevant information needed to identify the corner values. The convertor is necessary as these two classes don't serve the same purpose. I do think that `MDCCornerTreatment` is using outdated terminology, i.e., CutCorner doesn't exist anymore, it is AngledCorner.

Reply:
I see. I'm not entirely convinced yet that there needs to be two APIs that accomplish similar things with different terminology and naming. I see that you feel the two APIs accomplish very specific and different purposes, but this makes me wonder if we got the API wrong in the Shapes library and should update it accordingly. The end result being that MDCShapeCorner is part of the shapes library as a replacement for MDCCornerTreatment and the scheme relies on that accordingly.

If the above is sensible I think it's fair to plan a deprecation roadmap for the two APIs that will happen over a longer period of time, but that that work wouldn't necessarily have to block this PR.

Reply:
If we think we want this to be one entity, we can always add these properties to `MDCCornerTreatment` and the right initializers to it as well. I felt like the division kept it cleaner. If you feel strongly about it, i can do it in this PR, as this isn't a deprecation per-say.